exec/cnex/disassembly.c

```diff
@@ -0,0 +1,855 @@
+#include "disassembly.h"
+
+#include <stdio.h>
+#include <string.h>
+
+#include "bytecode.h"
+#include "exec.h"
+#include "module.h"
+#include "nstring.h"
+#include "opcode.h"
+
+#ifdef _MSC_VER
+#define snprintf _snprintf
+#endif
+
+
+static TInstructionDisassembler *g_disasm;
+
+TInstructionDisassembler *disasm_createDisassembler(void)
+{
+    TInstructionDisassembler *r = malloc(sizeof(TInstructionDisassembler));
+    if (r == NULL) {
+        fatal_error("Could not allocate memory for Instruction Dissembler.");
+    }
+    memset(r->out, 0x00, sizeof(r->out));
+    return r;
+}
+
+char *disasm_disassembleInstruction(TExecutor *exec)
+{
+    if (g_disasm == NULL) {
+        g_disasm = disasm_createDisassembler();
+    }
+
+    g_disasm->obj = exec->module->bytecode;
+    g_disasm->index = exec->ip;
+    g_disasm->out[0] = '\0';
+    disasm_disassemble(g_disasm);
+    return g_disasm->out;
+}
+
+void disasm_PUSHB(TInstructionDisassembler *disasm)
+{
+    BOOL val = disasm->obj->code[disasm->index+1] != 0;
+    disasm->index += 2;
+    sprintf(&disasm->out[0], "PUSHB %d", val);
+}
+
+void disasm_PUSHN(TInstructionDisassembler *disasm)
+{
+    disasm->index++;
+    uint32_t val = get_vint(disasm->obj->code, disasm->obj->codelen, &disasm->index);
+    sprintf(&disasm->out[0], "PUSHN %s", TCSTR(disasm->obj->strings[val]));
```

Comment:
Is this right? I don't really understand what the `TCSTR` macro does. Why is this different from what should be the same sort of thing on line 60?

Reply:
It is.  Line 60 is wrong.  Fixed line 60.